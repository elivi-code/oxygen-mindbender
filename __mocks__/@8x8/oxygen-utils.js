export * from '@8x8/oxygen-utils';

const useFontSize = jest.fn(() => 16);

export { useFontSize };
