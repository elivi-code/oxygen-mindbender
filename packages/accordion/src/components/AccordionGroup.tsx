import React, {
  useCallback,
  useEffect,
  useMemo,
  useRef,
  useState,
} from 'react';
import { v4 as uuid } from 'uuid';
import { getTestAttributes } from '@8x8/oxygen-config';

import { Container, Wrapper } from '../styled/AccordionGroup';
import { AccordionGroupInterface } from '../types';

function AccordionGroup({
  children,
  testId = 'ACCORDION_GROUP',
  initialActiveElementId = null,
  shouldCloseOnActiveClick = true,
  hasFixedHeight = false,
  onChange,
  expandTrigger,
  translations,
  ...rest
}: AccordionGroupInterface) {
  const [activeElementId, setActiveElementId] = useState(
    initialActiveElementId,
  );
  const containerRef = useRef(null);
  const headersRef = useRef(null);
  const [forcedContentHeight, setForcedContentHeight] = useState(0);

  const accordionItemsList = useMemo(
    () => (Array.isArray(children) ? children : [children]),
    [children],
  );

  const idsList: string[] = useMemo(
    () =>
      accordionItemsList.map(accordionItem => accordionItem.props.id || uuid()),
    [accordionItemsList],
  );

  const toggleActiveChange = useCallback(
    (id: string, isExpanded: boolean) => {
      const newActiveId = !isExpanded && shouldCloseOnActiveClick ? null : id;

      const canUpdate = onChange?.(newActiveId) ?? true;

      if (canUpdate) {
        setActiveElementId(newActiveId);
      }
    },
    [shouldCloseOnActiveClick, onChange, setActiveElementId],
  );

  useEffect(() => {
    const rawHeight =
      containerRef.current.clientHeight - headersRef.current.clientHeight;

    setForcedContentHeight(Math.abs(rawHeight));
  }, []);

  return (
    <Container
      {...getTestAttributes(testId)}
      hasFixedHeight={hasFixedHeight}
      ref={containerRef}
      {...rest}
    >
      <Wrapper ref={headersRef}>
        {accordionItemsList.map((accordionItem, index) =>
          React.cloneElement(accordionItem, {
            key: idsList[index],
            id: idsList[index],
            onChange: toggleActiveChange,
            isExpanded:
              (!hasFixedHeight || forcedContentHeight) &&
              idsList[index] === activeElementId,
            testId: accordionItem.props.testId || `${testId}_ITEM`,
            forcedHeight: hasFixedHeight ? forcedContentHeight : 0,
            shouldCloseOnActiveClick,
            isContentScrollable:
              accordionItem.props.isContentScrollable ?? hasFixedHeight,
            // Group-level props with individual accordion props taking precedence
            expandTrigger: accordionItem.props.expandTrigger ?? expandTrigger,
            translations: accordionItem.props.translations ?? translations,
          }),
        )}
      </Wrapper>
    </Container>
  );
}

export default AccordionGroup;
