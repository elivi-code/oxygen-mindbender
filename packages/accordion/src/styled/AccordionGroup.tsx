import styled from 'styled-components';
import { Theme } from '@8x8/oxygen-config';

const Container = styled.div<{ theme: Theme; hasFixedHeight: boolean }>`
  background: ${({ theme }) => theme.ui06};
  color: ${({ theme }) => theme.textColor01};
  ${({ theme }) => ({ ...theme.body01 })};
  box-sizing: border-box;
  display: flex;
  flex-direction: column;
  ${({ hasFixedHeight }) => hasFixedHeight && 'height: 100%'};
`;

const Wrapper = styled.div`
  display: flex;
  flex-direction: column;
`;

export { Container, Wrapper };
