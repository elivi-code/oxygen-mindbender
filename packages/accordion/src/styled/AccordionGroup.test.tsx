import React from 'react';
import { render } from '@testing-library/react';
import { OxygenProvider } from '@8x8/oxygen-config';

import { Container } from './AccordionGroup';

describe('AccordionGroup styled components', () => {
  it('should render Container', () => {
    const { container } = render(
      <OxygenProvider themeName="light">
        <Container hasFixedHeight />
      </OxygenProvider>,
    );

    expect(container.firstChild).toMatchSnapshot();
  });
});
