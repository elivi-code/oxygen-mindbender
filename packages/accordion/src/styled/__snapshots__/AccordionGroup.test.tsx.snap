// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccordionGroup styled components should render Container 1`] = `
.c0 {
  background: #FFFFFF;
  color: #292929;
  font-family: Inter,sans-serif;
  font-size: 0.875rem;
  line-height: 1.25rem;
  font-weight: 400;
  -webkit-letter-spacing: -0.006rem;
  -moz-letter-spacing: -0.006rem;
  -ms-letter-spacing: -0.006rem;
  letter-spacing: -0.006rem;
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  height: 100%;
}

<div
  class="c0"
/>
`;
